apps/arweave/src/ar_pool.erl

```diff
@@ -0,0 +1,764 @@
+
+-behaviour(gen_server).
+
+-export([start_link/0, is_client/0, get_current_sessions/0, get_jobs/1,
+		get_latest_output/0, cache_jobs/1, process_partial_solution/1,
+		post_partial_solution/1]).
+
+-export([init/1, handle_cast/2, handle_call/3, handle_info/2, terminate/2]).
+
+-include_lib("arweave/include/ar_config.hrl").
+-include_lib("arweave/include/ar_consensus.hrl").
+-include_lib("arweave/include/ar_mining.hrl").
+-include_lib("arweave/include/ar_pool.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+-record(state, {
+	%% The most recent keys come first.
+	session_keys = [],
+	%% Key => [{Output, StepNumber, PartitionUpperBound, Seed, Diff}, ...]
+	jobs_by_session_key = maps:new(),
+	request_pid_by_ref = maps:new()
+}).
+
+%%%===================================================================
+%%% Public interface.
+%%%===================================================================
+
+%% @doc Start the server.
+start_link() ->
+	gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%% @doc Return true if we are a pool client.
+is_client() ->
+	{ok, Config} = application:get_env(arweave, config),
+	Config#config.pool_client == true.
+
+%% @doc Return a list of up to two most recently cached VDF session keys.
+get_current_sessions() ->
+	gen_server:call(?MODULE, get_current_sessions, infinity).
+
+%% @doc Return a set of the most recent cached jobs.
+get_jobs(PrevOutput) ->
+	gen_server:call(?MODULE, {get_jobs, PrevOutput}, infinity).
+
+%% @doc Return the most recent cached VDF output. Return an empty binary if the
+%% cache is empty.
+get_latest_output() ->
+	gen_server:call(?MODULE, get_latest_output, infinity).
+
+%% @doc Cache the given jobs.
+cache_jobs(Jobs) ->
+	gen_server:cast(?MODULE, {cache_jobs, Jobs}).
+
+%% @doc Validate the given (partial) solution. If the solution is eligible for
+%% producing a block, produce and publish a block.
+process_partial_solution(Solution) ->
+	gen_server:call(?MODULE, {process_partial_solution, Solution}, infinity).
+
+%% @doc Send the partial solution to the pool.
+post_partial_solution(Solution) ->
+	gen_server:cast(?MODULE, {post_partial_solution, Solution}).
+
+%%%===================================================================
+%%% Generic server callbacks.
+%%%===================================================================
+
+init([]) ->
+	process_flag(trap_exit, true),
+	ok = ar_events:subscribe(solution),
+	{ok, #state{}}.
+
+handle_call(get_current_sessions, _From, State) ->
+	{reply, lists:sublist(State#state.session_keys, 2), State};
+
+handle_call({get_jobs, PrevOutput}, _From, State) ->
+	SessionKeys = State#state.session_keys,
+	JobCache = State#state.jobs_by_session_key,
+	{reply, get_jobs(PrevOutput, SessionKeys, JobCache), State};
+
+handle_call(get_latest_output, _From, State) ->
+	case State#state.session_keys of
+		[] ->
+			{reply, <<>>, State};
+		[Key | _] ->
+			{O, _SN, _U, _S, _Diff} = hd(maps:get(Key, State#state.jobs_by_session_key)),
```

Comment:
Rather than unpack and return the latest output. Could we remove this functino an djust have a `get_latest_job` that returns the latest `#job` record, and then the caller can access whatever field it needs?